### DISCUSSION TITLE
Precompile fails after adding tailwindcss-rails + npm modules

### DISCUSSION THREAD

**Discussion by sdhull** [fly-apps/dockerfile-rails#46] [Q&A]
Background: while my project is still young, I decided to switch from the "old & busted" bootstrap to "new hotness" Tailwind. Decided to use Flowbite so I don't have to build everything from scratch.

Anyway I added the [tailwindcss-rails](https://github.com/rails/tailwindcss-rails) gem and decided to use Node 18 + Yarn 3 (with `node_modules/`). So I did `yarn add flowbite` and `yarn add flowbite-typography`. Everything works great locally, even the `precompile` task runs fine.

I regenerated my Dockerfile with `rails g dockerfile`. The first problem I ran into was installing yarn. Apparently you can't `npm install -g yarn@3.6.1`, so I had to customize it in order to install Yarn 3
```
RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz -C /tmp/ && \
    /tmp/node-build-master/bin/node-build "${NODE_VERSION}" /usr/local/node && \
    corepack enable && \
    corepack prepare yarn@$YARN_VERSION --activate && \
    rm -rf /tmp/node-build-master
```
So that got me past that part. But now precompile is failing with:

```
 => ERROR [build 10/10] RUN SECRET_KEY_BASE=DUMMY ./bin/rails assets:precompile                                                                                              2.7s
------
 > [build 10/10] RUN SECRET_KEY_BASE=DUMMY ./bin/rails assets:precompile:
2.595
2.597 Rebuilding...
2.648 Error: Cannot find module 'flowbite/plugin'
2.648 Require stack:
2.648 - /rails/config/tailwind.config.js
2.648     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
2.648     at Function._resolveFilename (pkg/prelude/bootstrap.js:1955:46)
2.648     at Function.resolve (node:internal/modules/cjs/helpers:108:19)
2.648     at _resolve (/snapshot/tailwindcss/node_modules/jiti/dist/jiti.js:1:241025)
2.648     at jiti (/snapshot/tailwindcss/node_modules/jiti/dist/jiti.js:1:243309)
2.648     at /rails/config/tailwind.config.js:34:5
2.648     at jiti (/snapshot/tailwindcss/node_modules/jiti/dist/jiti.js:1:245784)
2.648     at /snapshot/tailwindcss/lib/lib/load-config.js:37:30
2.648     at loadConfig (/snapshot/tailwindcss/lib/lib/load-config.js:39:6)
2.648     at Object.loadConfig (/snapshot/tailwindcss/lib/cli/build/plugin.js:135:49) {
2.648   code: 'MODULE_NOT_FOUND',
2.648   requireStack: [ '/rails/config/tailwind.config.js' ]
2.648 }
```

I have zero idea how to debug this. I'd love to just be able to inspect the filesystem at this point. Why hasn't `docker` introduced a `debugger` line so I can inspect failing builds?? 😭

**Comment by sdhull** [comment]
OK update: I forgot I also added the `--cache` option because I wanted to speed up my builds. Also I copied the `Dockerfile` into a `Dockerfile.partial` and deleted the failing command and everything that came after so I could run the resulting image and poke around. I realized immediately that there was no `node_modules` directory.

I think the issue boils down to this:
```
# Install node modules
COPY --link package.json yarn.lock ./
RUN --mount=type=cache,id=bld-yarn-cache,target=/root/.yarn \
    YARN_CACHE_FOLDER=/root/.yarn yarn install --frozen-lockfile

# ... omitted other stuff

# Precompiling assets for production without requiring secret RAILS_MASTER_KEY
RUN SECRET_KEY_BASE=DUMMY ./bin/rails assets:precompile
```
At first I was thinking the cache mount was to blame, but after some experimentation, I realized that's not it. The issue is (apparently?) that `node_modules` gets clobbered by `COPY --link . .`

Re-ordering these so that `COPY` comes before yarn install seems to help but I'm a bit worried this is gonna invalidate all layers that come afterward on every deploy. Nonetheless, progress.

**Comment by sdhull** [comment]
I guess these lines are right next to each other, so it's just the `yarn install` that will have to happen on every deploy. But at least there's a persistent yarn cache 😎

**Comment by rubys** [comment]
The following code in the template should have installed yarn with corelib:

https://github.com/fly-apps/dockerfile-rails/blob/main/lib/generators/templates/_install_node.erb#L19-L30

If `node_modules` in your `.dockerignore` file?  Either way, I can't explain why it is empty after the `COPY` statement.

**Reply by sdhull** [reply]
Yes, it is... I didn't add it though, it was generated when I was getting fly.io deployment working 🤔
And to be fair, at that point I didn't have a `package.json` file or anything in `node_modules/` yet

**Reply by sdhull** [reply]
Should I remove it from `.dockerignore`?

**Comment by rubys** [comment]
If you used `fly launch` and didn't yet have a package.json file, then you weren't using dockerfile-rails.  That could explain the yarn install issues.

Yes, you want node_modules in your .dockerignore.

I'll try to reproduce the caching problem tomorrow.
